pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -722,6 +735,24 @@ protected WriteInEventLoopCallback newObject(Handle<WriteInEventLoopCallback> ha
 
     @Override
     public CompletableFuture<Void> closeAsync() {
+        return closeAsync(true);
+    }
+
+    public CompletableFuture<Void> closeAsync(boolean waitForPendingMessage) {
+        CompletableFuture<Void> closeResult = new CompletableFuture<>();
+        closeAsync(waitForPendingMessage, closeResult);
+        return closeResult;
+    }
+
+    private void closeAsync(boolean waitForPendingMessage, CompletableFuture<Void> closeFuture) {
```

Comment:
We shouldn't need the 2 new methods if we only care about the case `waitForPendingMessage=true`.